driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -153,119 +155,165 @@ private static String extractMessage(Throwable t) {
+        return new AsyncFunction<Message.Response, Void>() {
+            @Override
+            public ListenableFuture<Void> apply(Message.Response response) throws Exception {
+                switch (response.type) {
+                    case READY:
+                        return checkClusterName(executor);
+                    case ERROR:
+                        Responses.Error error = (Responses.Error)response;
+                        // Testing for a specific string is a tad fragile but well, we don't have much choice
+                        if (error.code == ExceptionCode.PROTOCOL_ERROR && error.message.contains("Invalid or unsupported protocol version"))
+                            throw unsupportedProtocolVersionException(protocolVersion);
+                        throw defunct(new TransportException(address, String.format("Error initializing connection: %s", error.message)));
+                    case AUTHENTICATE:
+                        Authenticator authenticator = factory.authProvider.newAuthenticator(address);
+                        if (protocolVersion == 1)
+                        {
+                            if (authenticator instanceof ProtocolV1Authenticator)
+                                return authenticateV1(authenticator, executor);
+                            else
+                                // DSE 3.x always uses SASL authentication backported from protocol v2
+                                return authenticateV2(authenticator, executor);
+                        }
+                        else
+                            return authenticateV2(authenticator, executor);
+                    default:
+                        throw defunct(new TransportException(address, String.format("Unexpected %s response message from server to a STARTUP message", response.type)));
+                }
+            }
+        };
     }
 
-    private UnsupportedProtocolVersionException unsupportedProtocolVersionException(int triedVersion) {
-        logger.debug("Got unsupported protocol version error from {} for version {}", address, triedVersion);
-        return new UnsupportedProtocolVersionException(address, triedVersion);
+    // Due to C* gossip bugs, system.peers may report nodes that are gone from the cluster.
+    // If these nodes have been recommissionned to another cluster and are up, nothing prevents the driver from connecting
+    // to them. So we check that the cluster the node thinks it belongs to is our cluster (JAVA-397).
+    private ListenableFuture<Void> checkClusterName(final ListeningExecutorService executor) {
+        final String expected = factory.manager.metadata.clusterName;
+
+        // At initialization, the cluster is not known yet
+        if (expected == null)
+            return MoreFutures.VOID_SUCCESS;
+
+        DefaultResultSetFuture clusterNameFuture = new DefaultResultSetFuture(null,new Requests.Query("select cluster_name from system.local"));
```

Comment:
Ran into a small bug while running a timeout stress test:

    Exception in thread "cluster1-worker-1" java.lang.NullPointerException
    	at com.datastax.driver.core.DefaultResultSetFuture.cancel(DefaultResultSetFuture.java:252)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:827)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:828)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:828)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:827)
    	at com.datastax.driver.core.HostConnectionPool.forceClose(HostConnectionPool.java:135)
    	at com.datastax.driver.core.HostConnectionPool.access$400(HostConnectionPool.java:45)
    	at com.datastax.driver.core.HostConnectionPool$2.onFailure(HostConnectionPool.java:122)
    	at com.google.common.util.concurrent.Futures$4.run(Futures.java:1140)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    	at java.lang.Thread.run(Thread.java:695)

I think this is because clusterNameFuture doesn't have a handler, so when a connection gets closed while being constructed and was in the process of checking cluster name this could happen.

I'm unsure of the implications, but since forceClose throws an exception as soon as it hits this exception, other connections may not be cancelled/scheduled for forceClose.  I think doing a simple null check on the handler may fix this, but testing to make sure that doesn't introduce any possible problems since handler is set after construction of DefaultResultSetFuture.